lib/iris/experimental/ugrid.py

```diff
@@ -1680,102 +1679,316 @@ def remove(
+        long_name=None,
+        var_name=None,
+        attributes=None,
+        cf_role=None,
+        node=None,
+        edge=None,
+        face=None,
+    ):
+        members = [c for c in self._members.values() if c is not None]
+
+        if cf_role is not None:
+            members = [
+                instance_
+                for instance_ in members
+                if instance_.cf_role == cf_role
+            ]
+
+        def location_filter(instances_, parameter_, location_name_):
+            if parameter_ is False:
+                members_ = [
+                    instance_
+                    for instance_ in instances_
+                    if location_name_
+                    not in (instance_.src_location, instance_.tgt_location)
+                ]
+            elif parameter_ is None:
+                members_ = instances_
+            else:
+                # Interpret any other value as =True.
+                members_ = [
+                    instance_
+                    for instance_ in instances_
+                    if location_name_
+                    in (instance_.src_location, instance_.tgt_location)
+                ]
+
+            return members_
+
+        for parameter, location_name in (
+            (node, "node"),
+            (edge, "edge"),
+            (face, "face"),
+        ):
+            members = location_filter(members, parameter, location_name)
+
+        # No need to actually modify filtering behaviour - already won't return
+        # any face cf-roles if none are present.
+        if face and self.NDIM < 2:
+            message = (
+                "Ignoring request to filter for non-existent 'face' cf-roles."
+            )
+            logger.debug(message, extra=dict(cls=self.__class__.__name__))
+
+        result = filter_cf(
+            members,
+            item=item,
+            standard_name=standard_name,
+            long_name=long_name,
+            var_name=var_name,
+            attributes=attributes,
+        )
+
+        # Use the results to filter the _members dict for returning.
+        result_ids = [id(r) for r in result]
+        result_dict = {
+            k: v for k, v in self._members.items() if id(v) in result_ids
+        }
+        return result_dict
+
+    def remove(
+        self,
+        item=None,
+        standard_name=None,
+        long_name=None,
+        var_name=None,
+        attributes=None,
+        cf_role=None,
+        node=None,
+        edge=None,
+        face=None,
+    ):
+        removal_dict = self.filters(
+            item=item,
+            standard_name=standard_name,
+            long_name=long_name,
+            var_name=var_name,
+            attributes=attributes,
+            cf_role=cf_role,
+            node=node,
+            edge=edge,
+            face=face,
+        )
+        for cf_role in self.REQUIRED:
+            not_removed = removal_dict.pop(cf_role, None)
+            if not_removed:
```

Comment:
@trexfeathers Minor. Had to stare at this for a while... kinda reads counter intuitively.

Perhaps would you consider renaming the `not_removed` variable? Just to help the developers that follow...

Reply:
80fb8e0401865fb1d0aff1d898585545490f6f82